### ISSUE TITLE
DeepSeekv3 CPU usage increases over time

### ISSUE THREAD

**Issue by kylesayrs** [vllm-project/llm-compressor#1650]
Continuation of https://github.com/vllm-project/llm-compressor/issues/1603#issuecomment-3034093072

**Comment by kylesayrs** [comment]
No smoking guns as of yet, but I have noticed some interesting patterns in terms of memory usage. I was isolate (at least part) of the memory increase to the forward passes of the model (as opposed to the GPTQ modifier).

For more context, I'm running the `examples/quantization_w4a16/llama3_example.py` example with `meta-llama/Llama-3.2-1B-Instruct` for testing purposes. I run a garbage collection step in between each sequential layer.

Here's a snapshot of the memory usage with a sequential pipeline running (with gptq)
<img width="1126" height="537" alt="Image" src="https://github.com/user-attachments/assets/6695eacf-7cef-45f5-ae56-50939e863c5f" />

And here's a snapshot of the memory usage with the sequential pipeline running (with no gptq)
<img width="1131" height="541" alt="Image" src="https://github.com/user-attachments/assets/7bf298e6-669d-41c1-b644-464d2fe3aa2b" />

Given that the memory increase over time is roughly the same for both (about 1.5Gb), the memory increase is likely due to the sequential pipeline itself, not the modifiers.

**Comment by kylesayrs** [comment]
I found there there is some interesting behavior if you run two pipelines back-to-back.

Here's a snapshot of the memory usage of running two sequential pipelines back-to-back (with no modifiers). The first increase is the first pipeline running, while the flat line after the peak is the second running (clipped).
<img width="1118" height="535" alt="Image" src="https://github.com/user-attachments/assets/f938182d-354f-4788-b862-38cd0b9dcf5f" />

Here's a snapshot of the memory usage of running a basic pipeline followed by a sequential pipeline (with no modifiers). Again, the second one is clipped.

<img width="1124" height="534" alt="Image" src="https://github.com/user-attachments/assets/1bb82a6a-921f-4ac5-88e5-5e3f37c96d50" />

EDIT: you'll notice that the peak memory usage of the basic pipeline is greater than that of the sequential pipeline. Also, basic -> seq causes the second seq run to have higher memory usage.

**Comment by kylesayrs** [comment]
This is the memory graph if you run two sequential pipelines, with the first sequential pipeline ingesting the entire dataloader into the IntermediatesCache, but only using one sample.

<img width="1127" height="537" alt="Image" src="https://github.com/user-attachments/assets/11184e20-9fa9-4a36-8e5d-24e5e3e68151" />

1. The initial memory increase as a result of running the sequential pipeline is roughly the same, whether you use one sample or the entire dataset (the increase is about 2gb)

**Comment by ramblingcoder** [comment]
@yongho-chang included since they originally opened the previous issue.

I assume by "clipped" you just mean the graph isn't for the full execution and just the important part of the second run?

I believe I more or less follow what you've outlined. Is there anything I can do on my side to help further the investigation?

**Comment by kylesayrs** [comment]
@ramblingcoder 

> I assume by "clipped" you just mean the graph isn't for the full execution and just the important part of the second run?

Here I'm just referring to how the screenshot is cropped to show the relevant bits, and doesn't show the full run

> I believe I more or less follow what you've outlined. Is there anything I can do on my side to help further the investigation?

Given that the sequential pipeline has a lower peak memory usage than the basic pipeline (despite starting at a lower value then increasing over time), this issue likely doesn't require an urgent fix.

I have a couple theories as to why memory consumption might increase as layers are executed ([kv cache](https://github.com/vllm-project/llm-compressor/blob/main/src/llmcompressor/modifiers/quantization/cache.py), [_can_record_outputs](https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L320C5-L320C24)), although these theories do not explain why subsequent passes would not also increase in memory usage over time.

Feel free to do some investigation for yourself. I've been using https://github.com/bloomberg/memray and https://github.com/kylesayrs/PTTP/tree/main/src/pttp to track memory usage.

EDIT: This issue has been re-prioritized on my end

**Comment by ramblingcoder** [comment]
Awesome, Appreciate the insights! Thank you for the additional details and how to troubleshoot it further.

Not familiar with those two tools so sounds like a fun thing to learn over the weekend. Will keep an eye on the issue in case anything comes up and appreciate your time.

**Comment by kylesayrs** [comment]
I've isolated this issue to something related to the `update_offload_parameter` function provided by compressed tensors. It seems that when this function is called, the original un-updated memory sticks around. Still debugging, will continue to post updates

**Comment by kylesayrs** [comment]
I was able to reduce heap memory usage by making a small tweak to the `update_offload_parameter` function. Instead of creating a new tensor, we instead copy data into the original tensor.

```
- module._hf_hook.weights_map.dataset.state_dict[name] = param.data.to("cpu")  # not actual code, but gets the point across
+ module._hf_hook.weights_map.dataset.state_dict[name].copy_(param.data)
```

However, I've checked the references to the original tensor using `gc.get_referrers` and found that it's only referred to by one parent, so I'm not sure why it wasn't being cleaned up (I also periodically call `gc.collect()`).

You can see that this change results in a no-longer-increasing heap memory size. However, the resident size still continues to grow, which is unfortunately the metric that we ultimately care about.

<img width="1128" height="539" alt="Image" src="https://github.com/user-attachments/assets/1a2fac9a-05bc-4147-b7d9-8686c829f1e4" />

EDIT: from the memory profiler, it seems like this increasing memory is due to this line in the activations cache.

```
else value.to(device=self.offload_device)
```

**Comment by brian-dellabetta** [comment]
> You can see that this change results in a no-longer-increasing heap memory size. However, the resident size still continues to grow, which is unfortunately the metric that we ultimately care about.

@kylesayrs this is my first hearing of "resident memory", but if it's the memory used by the process doesn't that include the heap? Preventing heap memory leaks ought to result in memory savings for users

**Comment by kylesayrs** [comment]
@brian-dellabetta Resident set size memory (RSS) is the physical memory used by RAM and [ultimately the metric that triggers OOM errors](https://groups.google.com/g/golang-nuts/c/dnRTLW3joPM).

The increase in RSS while heap stays constant could be due to a couple things
1. RSS is being lazily freed
2. Memory fragmentation is occurring
3. Some memory allocation is occurring that isn't captured by this profiler

So a little bit more work needs to be done to verify that it's not (2) or (3) before we can call success with (1). It's still very unclear to me how using `copy_` is more memory efficient where the previous code wasn't.

EDIT: I've done more investigation, and both of the tensors involved (the old tensor and the newly created tensor) only have one reference each. As I understand it, the explanation is that pytorch is holding on to memory it shouldn't, ie a pytorch memory bug.

**Comment by kylesayrs** [comment]
https://github.com/vllm-project/llm-compressor/pull/1670
https://github.com/neuralmagic/compressed-tensors/pull/399

**Comment by kylesayrs** [comment]
Not sure how I missed this earlier 🤦

https://github.com/huggingface/transformers/blob/9b7244f18941bfb9c3ebd5760e36a90722d17a6a/src/transformers/modeling_utils.py#L575-L583

**Comment by brian-dellabetta** [comment]
> Not sure how I missed this earlier 🤦
> 
> https://github.com/huggingface/transformers/blob/9b7244f18941bfb9c3ebd5760e36a90722d17a6a/src/transformers/modeling_utils.py#L575-L583

Not seeing why this is important, what am i missing?

**Comment by kylesayrs** [comment]
@brian-dellabetta When huggingface loads model weights, they take advantage of a OS feature called memory mapping (mmapping). This means that when weights are read, they are read from pointers which point to disk locations, not cpu locations. This means that a model does not need to load the entire model into CPU memory. However, when the model weights are written to (ie, when they are modified by quantization), they are written to CPU memory addresses. This explains why modifying a model layer-by-layer increases total CPU memory usage over time.

**Comment by brian-dellabetta** [comment]
@kylesayrs does that also imply the host cpu memory usage will be capped to the full model size? I would think this would be desired behavior to some extent?

**Comment by kylesayrs** [comment]
> does that also imply the host cpu memory usage will be capped to the full model size
@brian-dellabetta Yes, it does and this is the behavior we observe. This explanation only serves to explain why CPU memory usage starts lower than expected at the beginning.

Supporting quantizing models with less CPU memory usage is a separate discussion, but could be enabled through disk offloading or otherwise.

**Comment by ashgold** [comment]
Sounds awesome!

Regarding AWQ on DeepSeek-V3, I allocated 1.5TB of host memory and performed AWQ quantization, which was completed successfully.
The pod memory increased to approximately 1.3TB, so currently, 1.4TB of host memory is required. :(

**Comment by brian-dellabetta** [comment]
Hi @ashgold , thanks for the data point, much appreciated!

**Comment by zzc98** [comment]
> > does that also imply the host cpu memory usage will be capped to the full model size
> > [@brian-dellabetta](https://github.com/brian-dellabetta) Yes, it does and this is the behavior we observe. This explanation only serves to explain why CPU memory usage starts lower than expected at the beginning.
> 
> Supporting quantizing models with less CPU memory usage is a separate discussion, but could be enabled through disk offloading or otherwise.

It is hoped that quantization support with disk offloading can be supported. Currently, quantization models still require sufficient memory, which sometimes does not meet the conditions

**Comment by kylesayrs** [comment]
https://github.com/vllm-project/llm-compressor/issues/1650#issuecomment-3127875359

**Comment by ramblingcoder** [comment]
Awesome, mystery solved. Thank you for the explanation!
